src/content/reference/react-dom/components/form.md

```diff
@@ -0,0 +1,451 @@
+```
+
+```js actions.js hidden
+export async function search() {
+  return new Promise((_, reject) => reject())
+}
+```
+
+```json package.json hidden
+{
+  "dependencies": {
+    "react": "18.3.0-canary-6db7f4209-20231021",
+    "react-dom": "18.3.0-canary-6db7f4209-20231021",
+    "react-scripts": "^5.0.0",
+    "react-error-boundary": "4.0.3"
+  },
+  "main": "/index.js",
+  "devDependencies": {}
+}
+```
+
+</Sandpack>
+
+To surface error messages to your users from a form action you can use the `useFormState` Hook. `useFormState` takes two parameters: an action and a inital state, and returns two values, a state varible and a form action. The form action is passed to the `action` prop of the form. The value returned by the action passed to `useFormState` is used to update the state varible. This can be used to return error message that can be displayed to a user. When use in conjunction with [Server Actions](/reference/react/use-server) and [Server Components](/reference/react/use-client), the error message is progressively enhanced.
+
+<Sandpack>
+
+```js App.js
+import { useFormState } from "react-dom";
+import { signUpNewUser } from "./api";
+
+export default function Page() {
+  async function signup(prevState, formData) {
+    "use server";
+    const email = formData.get("email");
+    try {
+      await signUpNewUser(email);
+      alert(`Added "${email}"`);
+    } catch (err) {
+      return err.toString();
+    }
+  }
+  const [message, formAction] = useFormState(signup, null);
+  return (
+    <>
+      <h1>Signup for my newsletter</h1>
+      <p>Signup with the same email twice to see an error</p>
+      <form action={formAction} id="signup-form">
+        <label for="email">Email: </label>
+        <input name="email" id="email" placeholder="react@example.com" />
+        <button>Sign up</button>
+        {!!message && <p>{message}</p>}
+      </form>
+    </>
+  );
+}
+```
+
+```js api.js hidden
+let emails = [];
+
+export async function signUpNewUser(newEmail) {
+  if (emails.includes(newEmail)) {
+    throw new Error("This email address has already been added");
+  }
+  emails.push(newEmail);
+}
+```
+
+```json package.json hidden
+{
+  "dependencies": {
+    "react": "18.3.0-canary-6db7f4209-20231021",
+    "react-dom": "18.3.0-canary-6db7f4209-20231021",
+    "react-scripts": "^5.0.0"
+  },
+  "main": "/index.js",
+  "devDependencies": {}
+}
+```
+
+</Sandpack>
+
+### Processing forms differently with different submit buttons {/*processing-forms-differently-with-different-submit-buttons*/}
+
+Forms can be designed to handle multiple submission actions based on the button pressed by the user. Each button can be associated with a distinct action or behavior. When a user taps a specific button, the form is submitted, and a corresponding action, defined by that button's attributes and action, is executed. For instance, a form might allow users to either save their input as a draft or submit it for review with separate buttons.
```

Comment:
Could we clarify or namedrop that `formAction` allows you to customize the submission action for multiple buttons?

Reply:
Good catch, fixed.